Epic 8151 : WNPRC Purchasing System Module - Request Approval

labkey-bpatel opened the pull request:
#### Rationale
Workflow for the purchase admins to be able to review the request entered by the lab end users; and add, remove, or update line items, and any other inputs in the data entry form. Additionally, include new input fields specific to the purchase admin's process to update Assigned To, Status, and other admin related inputs.
Part of this change also includes addition of attachment form panel, where lab end users can upload jpg and pdf docs/packing slips.

#### Related Pull Requests
https://github.com/LabKey/ehrModules/pull/122
https://github.com/LabKey/labkey-ui-components/pull/442

#### Changes
- Add extensible columns upon folder creation.
- Add WNPRC Purchasing Schema & creditCardOptions table.
- Set filter on isValidVendor in the UI and sort dropdowns.
- Implement Purchasing Admin input panel with date columns for Order Date & Card Post date.
- Implement Form Attachment panel & handle save, and add link to the attachment.
- Configure request rowId to link to the existing request.
- Handle displaying of saved input fields and attachments.
- Add warning for 'Other' Account.
- Handle server side deletion of an existing line item row(s).
- Add calculated column 'Attachments' to ehr_purchasing.purchaseRequests table.
- Re-arrange panel rows and make styling adjustments.
- Handle errors on the server side (previously done client side).
- Update filtering on Accounts dropdown - only show users with account associated with that user.
- Handle permissions on the landing page - only show requests entered by the current user.
- Handle permissions on the data entry page - only purchase admins should see the form.

labkey-bpatel committed:
Add extensible columns upon folder creation.

labkey-bpatel committed:
Add WNPRC Purchasing schema & creditCardOptions table.

labkey-bpatel committed:
Filter on isValidVendor. Sort dropdowns.

labkey-bpatel committed:
Configure requestId to link request form in the overview report. Load request data entry form with data for specific requestId.

labkey-bpatel committed:
Handle save on an existing request.

labkey-bpatel committed:
Purchasing Admin panel

labkey-bpatel committed:
Right align number input

labkey-bpatel committed:
Adjust margin on Total

labkey-bpatel committed:
Remove requestId. Refactor to use rowId.

labkey-bpatel committed:
Purchasing admin panel input fields - adjustments & save.

labkey-bpatel committed:
Implement attachement panel and save. Update saving of existing request.

labkey-bpatel committed:
Remove 'Request Approved' data.

labkey-bpatel committed:
Load existing attachments in the attachment panel. Add extensible col 'attachments' in purchasingRequests, and handle save.

labkey-bpatel committed:
Handle display of existing attachment & removal from the UI. Format attachment files with url (wip).

labkey-bpatel committed:
Handle saved line items' row deletion.

labkey-bpatel committed:
Add warning for Other Account. Some cleanup

labkey-bpatel committed:
Subtotal & total related cleanup and alignment adjustments

labkey-bpatel committed:
Add link to the saved attachment in the data entry page. 'Other account' warning related fixes.

labkey-bpatel committed:
Update purchasingRequests metadata.

labkey-bpatel committed:
Calculated column 'Attachments' for purchasing requests table (displays attachments uploaded to the file server for a given request).

labkey-bpatel committed:
Re-arrange panel rows & styling adjustments

labkey-bpatel committed:
Add order date and card post date. Refactor around removal of attachments extensible column.

labkey-bpatel committed:
cleanup

labkey-bpatel committed:
Filter on user. Add link to "Request No." only if the user is a purchasing admin.

labkey-bpatel committed:
Add query to filter out accounts associated with the current user

labkey-bpatel committed:
Update error handling. Other refactor and cleanup around this change.

labkey-bpatel committed:
Update error handling on line items - move handling to server side.

labkey-bpatel committed:
Minor updates on error handling

labkey-bpatel committed:
Handle permissions on the data entry form - only have purchasing admin to update the existing request

labkey-bpatel committed:
cleanup

labkey-bpatel committed:
Permission related updates. Use of DatePicker instead of Date input. Cleanup.

labkey-bpatel committed:
Updates around using DatePicker. Add temporary re-route to ehr_purchasing.purchasingRequests grid view after submitting reviewed requests.

labkey-bpatel committed:
Fix loading weirdness around checking for permissions.

labkey-bpatel commented:
@labkey-jeckels - Please review non-react changes. Thank you!

labkey-bpatel commented:
@labkey-martyp - Please review react changes. Thank you!

labkey-bpatel committed:
Update versions

labkey-jeckels commented on WNPRC_Purchasing/resources/queries/ehr_purchasing/userAccountsAssocFiltered.sql:
```diff
@@ -0,0 +1,25 @@
+               WHERE ua.accessToAllAccounts = true
+                 AND ua.account IS NULL) u
+
+         UNION ALL
```
I think you want UNION instead of UNION ALL here in case a user is configured with double-access (both to individual accounts and all accounts).

labkey-bpatel replied:
Yup! now changed!

labkey-jeckels commented on WNPRC_Purchasing/src/org/labkey/wnprc_purchasing/WNPRC_PurchasingManager.java:
```diff
@@ -177,33 +219,171 @@ public void submitRequestForm(User user, Container container, WNPRC_PurchasingCo
+                    row.put("itemUnitId", lineItem.get("itemUnit"));
+
+                if (null == lineItem.get("unitCost"))
+                    lineItemErrors.addError(new PropertyValidationError("Required numeric value for 'Unit Cost' not provided", "unitCost"));
```
Are these checks needed? I think you could make sure the field's metadata knows it's required and the insertRows() and updateRows() should handle it.

labkey-bpatel replied:
It is not needed for the required fields indeed, but i added it explicitly because the errors from insertRows() and updateRows() were not as user friendly.

labkey-jeckels replied:
Is it mostly around wanting to use the label ("Unit Cost") instead of the column name ("unitCost")?

labkey-bpatel replied:
Mainly because the error was "Missing value for required property: unitCost" -- thought it was odd to see unitCost as "property" in the message/didn't think it was user friendly.

labkey-jeckels reviewed (approved):
Two suggestions, but no need for additional review

labkey-bpatel committed:
Code review change - use UNION instead of UNION ALL in user-account assoc. query.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_wnprc_purchasing_epic8151

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -15,26 +15,109 @@
+    const [permissionError, setPermissionError] = useState<string>();
 
     //equivalent to componentDidMount and componentDidUpdate (if with dependencies, then equivalent to componentDidUpdate)
     useEffect(() => {
```
This useEffect is really long, can you break some of this out into functions outside the component?

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -15,26 +15,109 @@
-        if (isNewRequest) {
+        const reqRowId = ActionURL.getParameter('requestRowId');
+        setRequestId(reqRowId);
+        Security.getUserPermissions({
```
I know we talked a little about this, I didn't realize the full context.  I believe this is already provided in getServerContext().user.  Either `getServerContext.user.isAdmin` or `getServerContext().user.get('permissionsList')`  so you shouldn't have to query the server here.

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -15,26 +15,109 @@
+                setRequestOrderModel(RequestOrderModel.create({
+                    rowId: vals[0].rowId,
+                    account: !!vals[0].otherAcctAndInves ? "Other" : vals[0].account,
+                    otherAcctAndInves: !!vals[0].otherAcctAndInves ? vals[0].otherAcctAndInves : undefined,
```
`vals[0].otherAcctAndInves || undefined`  recommend this

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -15,26 +15,109 @@
+            getData('ehr_purchasing', 'purchasingRequests', '*', undefined, filter).then(vals => {
+                setRequestOrderModel(RequestOrderModel.create({
+                    rowId: vals[0].rowId,
+                    account: !!vals[0].otherAcctAndInves ? "Other" : vals[0].account,
```
You don't need the bang-bang here.  The ternary is already evaluating the truthiness of the value on the left.  In general, bang-bang is not needed in javascript for conditional comparisons.  JS does all that behind the scenes.  The only real use for !! is if you want to convert a non-boolean value to a boolean to store in a property or something.

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -15,26 +15,109 @@
+            const filter = [Filter.create('rowId', reqRowId)];
+            getData('ehr_purchasing', 'purchasingRequests', '*', undefined, filter).then(vals => {
+                setRequestOrderModel(RequestOrderModel.create({
+                    rowId: vals[0].rowId,
```
It might be good to be defensive here and just verify vals is not empty before creating these models.  Otherwise you might get exception on direct access of vals

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -15,26 +15,109 @@
+                    confirmationNum: vals[0].confirmationNum,
+                    invoiceNum: vals[0].invoiceNum,
+                    orderDate: !!vals[0].orderDate ? new Date(vals[0].orderDate) : null,
+                    cardPostDate: !!vals[0].cardPostDate ? new Date(vals[0].cardPostDate) : null
```
Don't need !!

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -74,122 +167,215 @@ export const App : FC = memo(() => {
-        }
+        submitRequest(requestOrderModel,
+            lineItems,
+            !!requestId ? purchaseAdminModel : undefined,
```
!! not needed

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -74,122 +167,215 @@ export const App : FC = memo(() => {
+                if (r.success || r.fileNames?.length > 0)
+                {
+                    //TODO: this is temporary until purchasing admin page is created, then change it to get routed to the admin page
+                    if (!!requestId) {
```
!! not needed

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -74,122 +167,215 @@ export const App : FC = memo(() => {
+            (documentAttachmentModel.filesToUpload?.size > 0 || documentAttachmentModel.savedFiles?.length > 0) ? documentAttachmentModel : undefined,
+            lineItemRowsToDelete)
+            .then(r => {
+                if (r.success || r.fileNames?.length > 0)
```
Is `r.fileNames?.length > 0` a temporary condition until further work is done.  Otherwise it seems a little strange to look for success or that condition.

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -74,122 +167,215 @@ export const App : FC = memo(() => {
-        }
+                                // Errors for Request Order panel
+                                if (error.field === "account") {
+                                    requestOrderErrors.push({fieldName: 'account', errorMessage: error.msg || error.message});
```
`error.msg || error.message`  Is it possible to fix this in the submitRequest.api so there is a consistent message prop name?  If not, maybe get the error message at the beginning of this block (`const errorMessage = error.msg || error.message`) and use that for all these errorMessage properties.  You should be able to use shorthand if you use the same variable name as the property in that case `requestOrderErrors.push({fieldName: 'account', errorMessage})`

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -74,122 +167,215 @@ export const App : FC = memo(() => {
-            setLineItemErrorMsg(lineItemsErrorCount > 1 ? (msg + 's.') : (msg + '.'));
-        }
+                                // Errors for Request Order panel
+                                if (error.field === "account") {
```
Might be a good place for a switch-case statement

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -74,122 +167,215 @@ export const App : FC = memo(() => {
-                        onClick={onSaveBtnHandler}>Submit for Review
-                </button>
+                //has to be a purchasing admin to update the existing request
+                ((!!requestId && !hasPurchasingAdminPermission) || !!permissionError) &&
```
!! not needed

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -74,122 +167,215 @@ export const App : FC = memo(() => {
-            <RequestOrderPanel onInputChange={requestOrderModelChange} model={requestOrderModel}/>
-            <LineItemsPanel onChange={lineItemsChange} lineItems={lineItems} errorMsg={lineItemErrorMsg}/>
-            <button disabled={isSaving} className='btn btn-default' id='cancel' name='cancel' onClick={onCancelBtnHandler}>Cancel</button>
+        hasPermissionLoaded &&
```
If you don't need to load permissions, per my comment in the useEffect, then this shouldn't be necessary

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -74,122 +167,215 @@ export const App : FC = memo(() => {
-                    Saving...
-                </button>
+                //display ui for purchasing admins to update existing request or requesters to enter new request
+                ((!!requestId && hasPurchasingAdminPermission) || (requestId === undefined)) &&
```
!! no needed

labkey-martyp commented on WNPRC_Purchasing/src/client/RequestEntry/RequestEntry.tsx:
```diff
@@ -74,122 +167,215 @@ export const App : FC = memo(() => {
+                <>
+                    <RequestOrderPanel onInputChange={requestOrderModelChange} model={requestOrderModel}/>
+                    {
+                        !!requestId &&
```
You might not need !! here.  JSX is a little weird when the conditional is not as obvious

labkey-martyp commented on WNPRC_Purchasing/src/client/actions.ts:
```diff
@@ -1,38 +1,64 @@
+            sort: sort,
+            filterArray: filter,
+            success: function(results) {
+                if (results && results.rows) {
```
`results?.rows`

labkey-martyp commented on WNPRC_Purchasing/src/client/actions.ts:
```diff
@@ -1,38 +1,64 @@
+                qcState: purchasingAdminModel?.qcState || requestOrder.qcState,
+                assignedTo: purchasingAdminModel?.assignedTo,
+                creditCardOption: purchasingAdminModel?.creditCardOption,
+                program: !!purchasingAdminModel?.program ? purchasingAdminModel.program : '4',
```
!! not needed

labkey-martyp commented on WNPRC_Purchasing/src/client/actions.ts:
```diff
@@ -1,38 +1,64 @@
                 lineItems: lineItems,
-                hasNewVendor: (!!(requestOrder.vendor === 'Other' && VendorModel.getDisplayString(requestOrder.newVendor))),
+                lineItemsToDelete: lineItemsToDelete,
+                hasNewVendor: (!!(requestOrder.vendorId === 'Other' && VendorModel.getDisplayString(requestOrder.newVendor))),
```
!! not needed

labkey-martyp commented on WNPRC_Purchasing/src/client/actions.ts:
```diff
@@ -46,12 +72,67 @@ export async function submitRequest (requestOrder: RequestOrderModel, lineItems:
+            })
+            .catch(response => {
+                if (response) {
+                    const msg = 'Unable to load files in ' + (directory ? directory : 'root') + ': ' + response;
```
`` const msg = `Unable to load files in ${(directory ? directory : 'root')}: ${response}` ``   Template strings can be a little cleaner when strings are built with a lot of variables and logic.

labkey-martyp commented on WNPRC_Purchasing/src/client/components/DocumentAttachmentPanel.tsx:
```diff
@@ -0,0 +1,111 @@
+    //     });
+    //     onInputChange(updatedModel);
+    //
+    // }, [model, onInputChange]);
```
Remove commented out code?

labkey-martyp commented on WNPRC_Purchasing/src/client/components/DocumentAttachmentPanel.tsx:
```diff
@@ -0,0 +1,111 @@
+import {Draft, produce} from "immer";
+import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
+import {faTimesCircle} from "@fortawesome/free-solid-svg-icons";
+import {REMOVE_ATTACHMENT_ID_PREFIX} from "../constants";
```
Looks like some unused imports

labkey-martyp commented on WNPRC_Purchasing/src/client/components/DocumentAttachmentPanel.tsx:
```diff
@@ -0,0 +1,111 @@
+    //
+    // }, [model, onInputChange]);
+
+    const onFileRemove = useCallback((attachmentName: string) => {
```
onFileRemoval is optional on FileAttachmentForm.  I don't think you need this if you're not going to use it

labkey-martyp commented on WNPRC_Purchasing/src/client/components/DocumentAttachmentPanel.tsx:
```diff
@@ -0,0 +1,111 @@
+    }, [model, onInputChange]);
+
+    const showImg = useCallback((evt: any) => {
+        const attachment = model.savedFiles.filter(file => file.fileName === evt.target.innerText)
```
Could savedFiles be undefined?  `model.savedFiles?.filter`

labkey-martyp commented on WNPRC_Purchasing/src/client/components/DocumentAttachmentPanel.tsx:
```diff
@@ -0,0 +1,111 @@
+            const height = 400;
+            const y = window.top.outerHeight / 2 + window.top.screenY - ( height / 2);
+            const x = window.top.outerWidth / 2 + window.top.screenX - ( width / 2);
+            window.open(attachment[0].href,'popwin','width=${width}, height=${height}, top=${y}, left=${x}');
```
I'm not familiar with this, I'd like to see it when you have a chance.  Looks interesting

labkey-martyp commented on WNPRC_Purchasing/src/client/components/DocumentAttachmentPanel.tsx:
```diff
@@ -0,0 +1,111 @@
+                                            {/*     // onClick={onRemoveAttachment}*/}
+                                            {/*>*/}
+                                            {/*<FontAwesomeIcon className='fa-faTimesCircle' icon={faTimesCircle}/>*/}
+                                            {/*</span>*/}
```
Remove commented code?

labkey-martyp commented on WNPRC_Purchasing/src/client/components/RequestOrderPanel.tsx:
```diff
@@ -60,50 +60,109 @@ export const RequestOrderPanel: FC<Props> = memo((props) => {
+                {/*    value={model.shippingAttentionTo}*/}
+                {/*    hasError={model.errors?.find((field) => field.fieldName === 'shippingAttentionTo')}*/}
+                {/*    onChange={onValueChange}*/}
+                {/*/>*/}
```
Remove commented out code?

labkey-martyp reviewed (changes requested):
This looks great Binal.  Later we might consider if some of these inputs could be pulled into shared components for use elsewhere.  A lot of little comments mostly about syntax or cleanup.  We can go over anything that's not clear when you have a chance.

labkey-bpatel committed:
Bug fixes and code review changes:
- Make adjustments for smaller screen
- Use getServerContext().user instead of making an separate api call to check for permissions
- Add null checks
- Add switch case statement
- Use template string
- Remove commented code & unused imports
- Cleanup over usage of !!

labkey-bpatel committed:
Automated test: Update setup to include users to the folder, and to userAccountsAssociations. Add test plan/steps as log messages for the new test.

labkey-nicka committed:
Bump @labkey/components

labkey-nicka committed:
Rehydrate package-lock.json

labkey-bpatel committed:
Update @labkey/components to v.1.20.1

labkey-bpatel committed:
Lint formatting. Revert code review change for hasNewVendor check.

labkey-sweta committed:
Code review changes.

labkey-bpatel committed:
Assert permission text.

labkey-bpatel commented:
@labkey-sweta 
Test looks great! I've just made a minor update to assert permission text.
